sycl/plugins/level_zero/pi_level_zero.hpp

```diff
@@ -69,6 +70,12 @@ struct _pi_platform {
   // Cache versions info from zeDriverGetProperties.
   std::string ZeDriverVersion;
   std::string ZeDriverApiVersion;
+
+  // Cache pi_devices for reuse
+  std::vector<pi_device> PiDevicesCache;
+  std::mutex PiDevicesCacheMutex;
+  // Flag to indicate PiDevicesCache is invalidated
```

Comment:
Please add more info about what "invalid" cache really means? Why cannot we just clear the cache instead in such case?

Reply:
Whenever there is a call to piDeviceRelease, we invalidate the entire cache in pi_platform.
We don't physically clear the cache because it is going to be expensive to call the lower level L0 RT to get fresh set of ze_device_handle and create new pi_devices for them. 
Instead, we just reuse the cached pi_devices by calling initialize().
This is possible because I removed the call to destroy of initial command list in piDeviceRelease().

Reply:
Please add a TODO that we need to find a way to release/destroy the cache at tear-down. Currently the cahce seem to be valid forever after it is created, so do we really need the `CacheInvalidated` logic?

Reply:
I thought you said L0 RT will clean up all Ze_Device_Handle when the program ends.
If this is not true, yes we need to call L0 RT to clean up this memory.
Cache is valid as long as there is no call to piDeviceRelease() for it, although the memory for it is not de-allocated.
Yes, definititely we need CacheInvalided logic to tell the current cache is not valid and need to be re-initialized.
I will add a comment for tear-down cleanup necessity.

Reply:
Done

Reply:
>  it is going to be expensive to call the lower level L0 RT to get fresh set of ze_device_handle

L0 maintains a simialr cache themselves, so always same  L0 device handles are returned.
It sounds cheap to repeat the query for devices (compared to the subtle changes proposed here)

> Yes, definititely we need CacheInvalided logic to tell the current cache is not valid and need to be re-initialized.

Can't we instead not invalidate anything in piDeviceRelease, so that PI devices remain to be valid?

Reply:
> It sounds cheap to repeat the query for devices (compared to the subtle changes proposed here)

Still calling L0 RT is not as cheap as not calling it.


> Can't we instead not invalidate anything in piDeviceRelease, so that PI devices remain to be valid?

I am not sure if it is safe to keep using the pi_device without re-initializing them. Are you sure it is safe to use the pi_device without re-initialize them?

Reply:
I removed the invalidation logic per your request.